core/trino-main/src/main/java/io/trino/operator/StreamingExchangeClientBuffer.java

```diff
@@ -0,0 +1,224 @@
+import io.trino.execution.TaskId;
+import io.trino.execution.buffer.SerializedPage;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.Executor;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.util.concurrent.Futures.nonCancellationPropagating;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class StreamingExchangeClientBuffer
+        implements ExchangeClientBuffer
+{
+    private final Executor executor;
+    private final long bufferCapacityInBytes;
+
+    @GuardedBy("this")
+    private final Queue<SerializedPage> bufferedPages = new ArrayDeque<>();
+    @GuardedBy("this")
+    private volatile long bufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private volatile long maxBufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private volatile SettableFuture<Void> blocked = SettableFuture.create();
+    @GuardedBy("this")
+    private final Set<TaskId> activeTasks = new HashSet<>();
+    @GuardedBy("this")
+    private boolean noMoreTasks;
+    @GuardedBy("this")
+    private Throwable failure;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public StreamingExchangeClientBuffer(Executor executor, DataSize bufferCapacity)
+    {
+        this.executor = requireNonNull(executor, "executor is null");
+        this.bufferCapacityInBytes = requireNonNull(bufferCapacity, "bufferCapacity is null").toBytes();
+    }
+
+    @Override
+    public ListenableFuture<Void> isBlocked()
+    {
+        return nonCancellationPropagating(blocked);
+    }
+
+    @Override
+    public synchronized SerializedPage pollPage()
+    {
+        throwIfFailed();
+
+        if (closed) {
+            return null;
+        }
+        SerializedPage page = bufferedPages.poll();
+        if (page != null) {
+            bufferRetainedSizeInBytes -= page.getRetainedSizeInBytes();
+            checkState(bufferRetainedSizeInBytes >= 0, "unexpected bufferRetainedSizeInBytes: %s", bufferRetainedSizeInBytes);
+        }
+        // if buffer is empty block future calls
+        if (bufferedPages.isEmpty() && !isFinished() && blocked.isDone()) {
+            blocked = SettableFuture.create();
+        }
+        return page;
+    }
+
+    @Override
+    public synchronized void addTask(TaskId taskId)
+    {
+        if (closed) {
+            return;
+        }
+        checkState(!noMoreTasks, "no more tasks is expected");
+        activeTasks.add(taskId);
+    }
+
+    @Override
+    public void addPages(TaskId taskId, List<SerializedPage> pages)
+    {
+        long pagesRetainedSizeInBytes = 0;
+        for (SerializedPage page : pages) {
+            pagesRetainedSizeInBytes += page.getRetainedSizeInBytes();
+        }
```

Comment:
nit: stream map sum

Reply:
I also considered this method as a performance critical, thus was trying to avoid using streams.